lib/commands/exam.js

```diff
@@ -124,35 +180,98 @@ module.exports = TestCommand.extend({
-      var partitions = commandOptions.partition;
-
-      if (!partitions) {
-        partitions = [];
-        for (var i = 0; i < commandOptions.split; i++) {
-          partitions.push(i + 1);
-        }
-      }
+    const config = this._super._generateCustomConfigs.apply(this, arguments);
+    const customBaseUrl = this._getCustomBaseUrl(config, commandOptions);
 
-      if (Array.isArray(baseUrl)) {
-        var command = this;
-        config.testPage = baseUrl.reduce(function(testPages, baseUrl) {
-          return testPages.concat(command._generateTestPages(baseUrl, partitions, count));
-        }, []);
-      } else {
-        config.testPage = this._generateTestPages(baseUrl, partitions, count);
+    const shouldParallelize = this.validator.shouldParallelize;
+    const shouldLoadBalance = this.validator.shouldLoadBalance;
+    const shouldReplayExecution = this.validator.shouldReplayExecution;
+
+    if (!shouldParallelize && !shouldLoadBalance && !shouldReplayExecution) {
+      return config;
+    }
+
+    let browserIds = commandOptions.partition;
+    let appendingParam = '_partition';
+
+    if (shouldParallelize && !browserIds) {
+      browserIds = [];
+      for (var i = 0; i < commandOptions.split; i++) {
+        browserIds.push(i + 1);
       }
+    } else if (shouldLoadBalance) {
+      config.additional_browser_socket_events = this._addLoadBalancingBrowserSocketEvents(config);
+
+      const browserCount = commandOptions.loadBalance;
+      appendingParam = '_browser';
+
+      // Creates an array starting from 1 to browserCount. e.g. if brownserCount is 3, the returned array is [1, 2, 3]
+      browserIds = Array.from({length: browserCount}, (v,k) => k+1);
+    } else if (shouldReplayExecution) {
+      const executionMapping = this.getExecutionMappigInstance(commandOptions);
+      config.browser_module_mapping = executionMapping['executionMapping'];
+
+      appendingParam = '_browser';
+      browserIds = commandOptions.replayBrowser;
+    }
+
+    if (Array.isArray(customBaseUrl)) {
+      const command = this;
+      config.testPage = customBaseUrl.reduce(function(testPages, customBaseUrl) {
+        return testPages.concat(command._generateTestPages(customBaseUrl, appendingParam, browserIds));
+      }, []);
+    } else {
+      config.testPage = this._generateTestPages(customBaseUrl, appendingParam, browserIds);
     }
 
     return config;
   },
 
+  /**
+   * Customizes the base url by specified test splitting options - parellel or loadBalance.
+   * 
+   * @param {String} config
+   * @param {Object} commandOptions
+   * @return {Object}
+   * @example tests/index.html?hidepassed&_split=3 if parallel
+   *          tests/index.html?hidepassed&_split=3&_loadBalance if load-balance
+   *          tests/index.html?hidepassed&_split=3&_loadBalance&_partition=1&_partition=2 if load-balance and partitions are specified
+   */
+  _getCustomBaseUrl: function(config, commandOptions) {
```

Comment:
this seems a tad complicated, can we unit test?

Reply:
`getCustomBaseUrl` is used in `generateCustomConfigs` and we do have unit tests for `generateCustomConfigs`. do you still think that unit tests for  `getCustomBaseUrl` would be helpful? fyi the unit tests are in `node-tests/unit/commands/exam-test.js`
https://github.com/ember-cli/ember-exam/pull/136/files#r229002197

Reply:
Having a bit of redundancy in tests is OK imo, as it helps prevent regressions that happen as a result of other changes in the system. I'd add tests.

Reply:
Added tests but it will fail until #148 landed. It verifies to have ember-cli@3.2 above for load-balance option. Since in this pr ember-cli is lower version tests that does option validator for load-balance will be failed.